Sources/PriorityQueueModule/PriorityQueue.swift

```diff
@@ -0,0 +1,537 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Collections open source project
+//
+// Copyright (c) 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+import Swift
+
+/// A double-ended priority queue built on top of a [Min-Max Heap](https://en.wikipedia.org/wiki/Min-max_heap)
+/// data structure.
+///
+/// In a min-max heap, each node at an even level in the tree is less than all
+/// its descendants, while each node at an odd level in the tree is greater than
+/// all of its descendants.
+///
+/// The implementation is based off [Atkinson et al. Min-Max Heaps and
+/// Generalized Priority Queues (1986)](http://akira.ruc.dk/~keld/teaching/algoritmedesign_f03/Artikler/02/Atkinson86.pdf).
+///
+/// M.D. Atkinson, J.-R. Sack, N. Santoro, T. Strothotte. October 1986.
+/// Min-Max Heaps and Generalized Priority Queues. Communications of the ACM.
+/// 29(10):996-1000.
+public struct PriorityQueue<Element: Comparable> {
+  @usableFromInline
+  internal var storage: [Element]
+
+  /// A Boolean value indicating whether or not the queue is empty.
+  ///
+  /// - Complexity: O(1)
+  @inlinable
+  public var isEmpty: Bool {
+    storage.isEmpty
+  }
+
+  /// The number of elements in the queue.
+  ///
+  /// - Complexity: O(1)
+  @inlinable
+  public var count: Int {
+    storage.count
+  }
+
+  /// A read-only view into the underlying heap.
+  ///
+  /// In the current implementation, the elements aren't _arbitrarily_ ordered,
+  /// as a min-max heap is used for storage. However, no guarantees are given as
+  /// to the ordering of the elements.
+  ///
+  /// - Complexity: O(1)
+  @inlinable
+  public var unordered: [Element] {
+    storage
+  }
+
+  /// Creates an empty queue.
+  @inlinable
+  public init() {
+    storage = []
+  }
+
+  /// Inserts the given element into the queue.
+  ///
+  /// - Complexity: O(log `count`) / 2
```

Comment:
I think the complexity is O(log `count`) / 4 instead of / 2, since other than the first comparison, we skip the parent node to compare with the grandparent.

Reply:
My mistake. / 2 is correct. Please disregard the comment above.